File: app/src/main/java/me/zhanghai/android/files/provider/common/ByteStringBuilder.java
/*
 * Copyright (c) 2019 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.provider.common;

import java.util.Arrays;

import androidx.annotation.NonNull;

public class ByteStringBuilder {

    @NonNull
    private byte[] mBytes;
    private int mLength;

    public ByteStringBuilder(int capacity) {
        mBytes = new byte[capacity];
    }

    public ByteStringBuilder() {
        this(16);
    }

    public ByteStringBuilder(@NonNull ByteString byteString) {
        this(byteString.length() + 16);

        append(byteString);
    }

    public byte byteAt(int index) {
        if (!(index >= 0 && index < mLength)) {
            throw new IndexOutOfBoundsException();
        }
        return mBytes[index];
    }

    public int capacity() {
        return mBytes.length;
    }

    public int length() {
        return mLength;
    }

    public boolean isEmpty() {
        return mLength == 0;
    }

    @NonNull
    public ByteStringBuilder append(byte b) {
        ensureCapacity(mLength + 1);
        mBytes[mLength] = b;
        ++mLength;
        return this;
    }

    @NonNull
    public ByteStringBuilder append(@NonNull ByteString byteString) {
        int byteStringLength = byteString.length();
        ensureCapacity(mLength + byteStringLength);
        System.arraycopy(byteString.getOwnedBytes(), 0, mBytes, mLength, byteStringLength);
        mLength += byteStringLength;
        return this;
    }

    private void ensureCapacity(int minimumCapacity) {
        int capacity = mBytes.length;
        if (minimumCapacity > capacity) {
            int newCapacity = (capacity << 1) + 2;
            if (newCapacity < minimumCapacity) {
                newCapacity = minimumCapacity;
            }
            mBytes = Arrays.copyOf(mBytes, newCapacity);
        }
    }

    @NonNull
    public ByteString toByteString() {
        return ByteString.ofBytes(mBytes, 0, mLength);
    }

    /**
     * @deprecated Use {@link #toByteString()} instead.
     */
    @NonNull
    @Override
    public String toString() {
        return new String(mBytes, 0, mLength);
    }
}

File: app/src/main/java/me/zhanghai/android/files/provider/common/ByteStringBuilder.kt
/*
 * Copyright (c) 2019 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.provider.common

class ByteStringBuilder(capacity: Int = 16) {
    private var bytes = ByteArray(capacity)

    var length = 0
        private set

    constructor(byteString: ByteString) : this(byteString.length + 16) {
        append(byteString)
    }

    operator fun get(index: Int): Byte {
        if (index !in 0..<length) {
            throw IndexOutOfBoundsException()
        }
        return bytes[index]
    }

    val isEmpty: Boolean
        get() = length == 0

    fun capacity(): Int = bytes.size

    fun append(byte: Byte): ByteStringBuilder {
        ensureCapacity(length + 1)
        bytes[length] = byte
        ++length
        return this
    }

    fun append(bytes: ByteArray, start: Int = 0, end: Int = bytes.size): ByteStringBuilder {
        val newLength = length + (end - start)
        ensureCapacity(newLength)
        bytes.copyInto(this.bytes, length, start, end)
        length = newLength
        return this
    }

    fun append(byteString: ByteString): ByteStringBuilder = append(byteString.borrowBytes())

    private fun ensureCapacity(minimumCapacity: Int) {
        val capacity = bytes.size
        if (minimumCapacity > capacity) {
            var newCapacity = (capacity shl 1) + 2
            if (newCapacity < minimumCapacity) {
                newCapacity = minimumCapacity
            }
            bytes = bytes.copyOf(newCapacity)
        }
    }

    fun toByteString(): ByteString = bytes.toByteString(0, length)

    override fun toString(): String = String(bytes, 0, length)
}
